File: proto/xmlrpc/mylibrary.rb
class RubyLibraryExample

  def do_nothing()
  end
  
  def failure(message)
    raise message
  end
  
  def logging(message, level='INFO')
    puts '*'+ level + '* ' + message
  end

  def one_argument(arg)
    puts 'arg: ' + arg
  end

  def two_arguments(arg1, arg2)
    puts 'arg1: ' + arg1
    puts 'arg2: ' + arg2
  end

  def arguments_with_default_values(arg1, arg2='default value', arg3=nil)
    puts 'arg1: ' + arg1
    puts 'arg2: ' + arg2
    puts 'arg3: ' + arg3
  end

  def argument_should_be_boolean_true(arg)
    argument_type_should_be(arg, TrueClass)
  end
  
  def argument_should_be_string(arg)
    argument_type_should_be(arg, String)
  end

  def argument_type_should_be(arg, type)
    if not arg.class == type
      raise Exception, 'Argument type should be '+ type.to_s() +' but was ' + arg.class.to_s()
    end    
  end

  def return_string()
    return 'Hello, world!'
  end
  
  def return_object()
    return MyObjectToReturn.new
  end

  def return_true()
    return true
  end

  def return_multiple_values(given)
    return 'first', 2, given
  end
  
  def divide(a, b)
    return Float(a) / Float(b)
  end
    
end


class MyObjectToReturn
  def to_s()
    return "String representation of MyObject"
  end
end


if ARGV.size == 1
  require "robotxmlrpcserver"
  RobotXmlRpcServer.new(RubyLibraryExample.new, ARGV[0])
else
  puts "Usage: #{__FILE__} port"
end

exit

File: proto/xmlrpc/MyLibrary.py
import os.path
import sys
import types

class MyLibrary:
  
    def divide(self, a, b):
        return a / b
  
    def raise_failure(self, message):
        raise AssertionError(message)
  
    def logging(self, message, message_level='INFO'):
        print '*' + message_level + '* ' + message
  
    def return_nothing(self):
        return None
      
    def return_multiple_values(self, value1, value2):
        return value1, value2
  
    def check_argument_is_boolean_type_true(self, arg):
        self.arguments_type_should_be(arg, True.__class__)
        return arg
  
    def return_true():
        return True

    def check_argument_is_string_type(self, arg):
        self.arguments_type_should_be(arg, types.StringType)
        return arg

    def arguments_type_should_be(self, arg, type):
        if not arg.__class__ == type:
            raise Exception, 'Arguments type should be '+ type +' but was ' + arg.__class__
  
    def should_be_list(self, arg):
        if arg != ['a','b','c']:
            raise AsserionError, "Given list is not ['a', 'b', 'c']"
        return arg
  
    def should_be_dictionary(self, arg):
        if arg != {'a':1, 'b':'Hello', 'c':['a', 1]}:
            raise AssertionError, "Given argument is not {'a'=> 1, 'b'=>'Hello', 'c' => ['a', 1]}"
        return arg
  
    def return_object(self):
        return  MyObject()


class MyObject:
    def __str__(self):
        return "String representation of MyObject"

#METHODS FOR STARTING THE LIBRARY FROM COMMAND LINE

def start(port):
    sys.path.append(os.path.abspath(os.path.dirname(__file__)))
    from RobotXMLRPCServer import RobotXMLRPCServer
    
    server = RobotXMLRPCServer(MyLibrary(), port)
    server.register_function(server.get_keyword_names)
    server.register_function(server.run_keyword)
    server.register_introspection_functions()
    server.serve_forever()

def help():
    print """
  Usage: %s port"
""" % (__file__)

if __name__ == '__main__':
    if len(sys.argv) != 2:
        help()
    else:
        start(int(sys.argv[1]))
